src/Microsoft.ML.Core/Utilities/DoubleParser.cs

```diff
@@ -160,15 +164,15 @@ public static Result Parse(out Double value, string s, int ichMin, int ichLim)
             // Make sure everything was consumed.
             while (ichEnd < ichLim)
             {
-                if (!char.IsWhiteSpace(s[ichEnd]))
+                if (!char.IsWhiteSpace(s.Span[ichEnd]))
                     return Result.Extra;
                 ichEnd++;
             }
 
             return Result.Good;
         }
 
-        public static bool TryParse(out Single value, string s, int ichMin, int ichLim, out int ichEnd)
+        public static bool TryParse(out Single value, ReadOnlyMemory<char> s, int ichMin, int ichLim, out int ichEnd)
```

Comment:
>public static bool TryParse(out Single value, ReadOnlyMemory<char> s, int ichMin, int ichLim, out int ichEnd) [](start = 8, length = 109)

So regarding this:

`ReadOnlyMemory<char> s, int ichMin, int ichLim`

Why was it written that way? Couldn't the caller just slice the memory *themselves* rather than having us pass around ranges, with the necessary complexities that this entails? Isn't that simplification much of the reason we're doing this work in the first place?

This signature can then be simplifed to:

```csharp
public static bool TryParse(out Single value, ReadOnlyMemory<char> s, out int ichEnd)
```

Reply:
I think though, you *probably* want all these parse methods to be over `ReadOnlySpan<char>`, rather than `ReadOnlyMemory<char>`.

---
In reply to: [216127480](https://github.com/dotnet/machinelearning/pull/863#discussion_r216127480) [](ancestors = 216127480)

Reply:
100% with you!

---
In reply to: [216127512](https://github.com/dotnet/machinelearning/pull/863#discussion_r216127512) [](ancestors = 216127512,216127480)